plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -823,22 +829,91 @@ compileExpr e = traceCompilation 2 ("Compiling expr:" GHC.<+> GHC.ppr e) $ do
         body' <- compileExpr body
         pure $ PIR.mkLet annMayInline PIR.Rec binds body'
 
+    GHC.Case scrutinee b t alts ->
+      compileCase (const . GHC.isDeadOcc . GHC.occInfo . GHC.idInfo) True binfo scrutinee b t alts
+
+    -- we can use source notes to get a better context for the inner expression
+    -- these are put in when you compile with -g
+    -- See Note [What source locations to cover]
+    GHC.Tick tick body | Just src <- getSourceSpan maybeModBreaks tick ->
+      traceCompilation 1 ("Compiling expr at:" GHC.<+> GHC.ppr src) $ do
+        CompileContext{ccOpts = coverageOpts} <- ask
+        -- See Note [Coverage annotations]
+        let anns = Set.toList $ activeCoverageTypes coverageOpts
+        compiledBody <- fmap (addSrcSpan $ src ^. srcSpanIso) <$> compileExpr body
+        foldM (coverageCompile body (GHC.exprType body) src) compiledBody anns
+
+    -- ignore other annotations
+    GHC.Tick _ body -> compileExpr body
+    -- See Note [Coercions and newtypes]
+    GHC.Cast body _ -> compileExpr body
+    GHC.Type _ -> throwPlain $ UnsupportedError "Types as standalone expressions"
+    GHC.Coercion _ -> throwPlain $ UnsupportedError "Coercions as expressions"
+
+compileCase ::
+  (CompilingDefault uni fun m ann) =>
+  -- | Whether the variable is dead in the expr
+  (GHC.Var -> GHC.CoreExpr -> Bool) ->
+  -- | Whether we should try to rewrite opaque constructor applications
+  Bool ->
+  BuiltinsInfo uni fun ->
+  GHC.CoreExpr ->
+  GHC.Var ->
+  GHC.Type ->
+  [GHC.CoreAlt] ->
+  m (PIRTerm uni fun)
+compileCase isDead rewriteOpaque binfo scrutinee binder t alts = case alts of
+  [GHC.Alt con bs body]
     -- See Note [Evaluation-only cases]
-    GHC.Case scrutinee b _ [GHC.Alt _ bs body] | all (GHC.isDeadOcc . GHC.occInfo . GHC.idInfo) bs -> do
+    | all (`isDead` body) bs -> do
       -- See Note [At patterns]
       scrutinee' <- compileExpr scrutinee
-      withVarScoped b $ \v -> do
+      withVarScoped binder $ \v -> do
         body' <- compileExpr body
         -- See Note [At patterns]
         let binds = [PIR.TermBind annMayInline PIR.Strict v scrutinee']
         pure $ PIR.mkLet annMayInline PIR.NonRec binds body'
-    GHC.Case scrutinee b t alts -> do
+    | rewriteOpaque
+    , GHC.DataAlt dataCon <- con
+    , let dataConName = GHC.dataConName dataCon
+    , let dataConModule = GHC.moduleNameString . GHC.moduleName <$> GHC.nameModule_maybe dataConName
+    , let dataConBase = GHC.occNameString (GHC.nameOccName dataConName)
+    , (dataConModule, dataConBase) `elem` opaqueTypes -> do
```

Comment:
AFAICT, this transformation is totally safe regardless of what data constructor we use it on. If we see this pattern with a non-opaque data constructor, that's still pointless reboxing.

So... maybe we should just drop the whole "is this an opaque data constructor business". The downside is that then I guess this would fire more frequently.

Reply:
Good point. Will try that.

Reply:
Done; it actually led to some small performance wins!